src/nbl/video/IDescriptorPool.cpp

```diff
@@ -4,60 +4,63 @@
 namespace nbl::video
 {
 
-core::smart_refctd_ptr<IGPUDescriptorSet> IDescriptorPool::createDescriptorSet(core::smart_refctd_ptr<const IGPUDescriptorSetLayout>&& layout)
+bool IDescriptorPool::createDescriptorSets(uint32_t count, const IGPUDescriptorSetLayout* const* layouts, core::smart_refctd_ptr<IGPUDescriptorSet>* output)
 {
-    if (!isCompatibleDevicewise(layout.get()))
-        return nullptr;
+    core::vector<SDescriptorOffsets> descriptorOffsets(count);
 
-    return createDescriptorSet_impl(std::move(layout));
+    for (uint32_t i = 0u; i < count; ++i)
+    {
+        if (!isCompatibleDevicewise(layouts[i]))
+            return false;
```

Comment:
you need to "cancel" the preceeding successful allocations if you hit this

Or SIMPLER, check this first before the `core::vector` is constructed on the stack

Reply:
actually `allocateDescriptorOffsets` itself might fail as well.

I suggest rewriting the loop to this
```cpp
core::vector<SDescriptorOffsets> descriptorOffsets;
descriptorOffsets.reserve(count);

for (uint32_t i = 0u; i < count; ++i)
{
        if (!isCompatibleDevicewise(layouts[i]))
            break;
            
        if (!allocateDescriptorOffsets(descriptorOffsets.emplace_back(),layouts[i]))
        {
            descriptorOffsets.pop_back();
            break;
        }
}

const auto successCount = descriptorOffsets.size();
std::fill_n(output+successCount,count-successCount,nullptr);
return createDescriptorSets_impl(successCount,layouts,descriptorOffsets.data(),output);
```
changing the `allocateDescriptorOffsets` to return a boolean whether it suceeded.

Reply:
Resolved.

Reply:
Actually you'll need to be able to free the `descriptorOffsets` if `createDescriptorSets_impl` doesn't succeed, so we should be looking at
```cpp
while (descriptorOffsets.size()<count)
{
        const auto i = descriptorOffsets.size();
        if (!isCompatibleDevicewise(layouts[i]))
            break;
            
        if (!allocateDescriptorOffsets(descriptorOffsets.emplace_back(),layouts[i]))
        {
            descriptorOffsets.pop_back();
            break;
        }
}

const bool creationSuccess =  createDescriptorSets_impl(descriptorOffsets.size(),layouts,descriptorOffsets.data(),output);
const auto successCount = creationSuccess ? descriptorOffsets.size():0u;
std::fill_n(output+successCount,count-successCount,nullptr);
if (!creationSuccess)
    cancelDescriptorOffsetsAllocations(std::move(descriptorOffsets));
return successCount;
```

Reply:
Okay, I didn't think descriptor allocation failure from the pool's storage would be a common occurrence that's why I just put an `assert` there https://github.com/Devsh-Graphics-Programming/Nabla/pull/345#discussion_r1116361347.

Now that we're handling failure cases properly, are you sure we should be doing the following?
```C++
if (!isCompatibleDevicewise(layouts[i]))
    break;
```
If `layouts[i]` is not compatible we can just skip that layout and continue on with the rest that may or may not be valid. This will mean that `output` will be sparse --only having non `nullptr` values for compatible DS layouts. If we do this then we need to modify `createDescriptorSets_impl` because it uses `vkAllocateDescriptorSets` which works on contiguous ranges as opposed to `output` which could be sparse.

Reply:
I don't think we should be adding this much extra code to "handle" what is effectively API misuse.

Never assume the user is an idiot, and lets not code for one.

Reply:
Resolved.